src/v2/git_url.rs

```diff
@@ -42,7 +42,9 @@ impl GitUrl {
     pub fn new<S: Into<String>>(url: S) -> Result<GitUrl> {
         let url = url.into();
         if GitUrl::should_treat_as_url(&url) {
-            Ok(GitUrl { url: url })
+            let git_url = GitUrl { url };
+            git_url.parse_parts()?;
```

Comment:
Let's add a comment explaining the invariant we're trying to enforce (otherwise this might be a bit mysterious):

```rust
// Invariant: `parse_parts` should succeed for all `GitUrl` values after construction.
```